Add TPMKMS support

hslatman opened the pull request:
This PR adds support for using the TPMKMS with `step-kms-plugin`

Attestations for the `tpm` format are created in a slightly different manner than the existing formats. Instead of signing data with the private key, the certification of a key is done at key creation time. The certification facts are recorded in the TPMKMS, so that they can be used by other applications. In this case, `step kms attest --format tpm 'tpmkms:name=keyName'` will return an attestation staement that includes these key certification parameters.

Will do some more testing and add examples in follow up commits.

maraino commented on cmd/attest.go:
```diff
@@ -230,7 +266,8 @@ func init() {
-	flags.Var(format, "format", "The `format` to print the attestation.\nOptions are step or packed")
+	format := flagutil.LowerValue("format", []string{"", "step", "packed", "tpm"}, "")
+	flags.Var(format, "format", "The `format` to print the attestation.\nOptions are step, packed or tpm")
+	flags.Bool("bundle", false, "Print all certificates in the chain")
```
Doesn't it make more sense to default to bundled certificates? It will also be backward compatible.

What is the use case for showing only the leaf?

maraino commented on cmd/certificate.go:
```diff
@@ -116,4 +175,5 @@ func init() {
 	flags.SortFlags = false
 
 	flags.String("import", "", "The certificate `file` to import")
+	flags.Bool("bundle", false, "Print all certificates in the chain/bundle")
```
I would print the bundled certificate by default. A different option can be added to show only the leaf. Probably something cleaner that --bundled=false

hslatman replied:
I've left this as is, to be consistent with our other tooling.

maraino replied:
`step-kms-plugin certificate` is mainly used to load a certificate from a KMS when `step ca certificate --x5c-cert` is used. The `--x5c-cert` flag expects a bundled certificate, that is what you will use with files. The flag `--x5c-chain` was added to support a bundled certificate in those KMSs that only support a single certificate. There is some inconsistency because not all KMSs support the same things, but I don't think we want to add `--x5c-chain` unless it is necessary.

We can keep the flag here, for "consistency" with `step certificate inspect`, but we will need to change the cli to add the `--bundle` option. This will cause some incompatibility depending if you are using the last version of the plugin or not.

maraino reviewed (changes requested):
Looks good, but I would bundle by default, it is what we were doing before.

hslatman commented:
> Looks good, but I would bundle by default, it is what we were doing before.

Is it, though? There was only (Load|Store)Certificate, which would work with / return a single certificate? I may have missed something, but that was the reason to add the CertificateChainManager. So now, when the KMS supports chains, the default is to show just one cert in the output.

maraino commented:
The other KMSs only support one, but for example in attest we show the bundle.

hslatman commented:
Now I get what you mean! Yes, the behavior changed slightly for the YubiKey, due to adding the leaf as the first cert of the chain, and not looping through the entire chain now.

I think for `step kms certificate` returning a single certificate by default makes most sense. It would be similar to our `step certificate inspect`, which will print the info for a single cert only, unless `--bundle` is supplied. It doesn't say `step kms chain`, after all. 

For attestation use cases, I think most users will need the chain by default, so I agree that we may want the behavior of `--bundle` there by default. I could add `--leaf` to show just the leaf? If someone were to run `step kms attest .... | step certificate inspect`, they would need to add `--bundle` to show the full information, though.

dwmw2 commented:
Not entirely sure if this is the correct place for this "informational note": 

For TPM wrapped keys, there is a file format defined at https://www.hansenpartnership.com/draft-bottomley-tpm2-keys.html _(PEM files containing `-----BEGIN TSS2 PRIVATE KEY-----`)_ which is supported by other tools including the two OpenSSL engines, GnuTLS and OpenConnect.

The vision is that any software which can use keys from a file, should automatically recognise when the file they're given contains such a key (in either PEM or DER form) and Do The Right Thing™.

Tools which create keys to be used with a TPM should also use this storage format or at least be able to export to it.

hslatman commented:
@dwmw2 thank you for your insight! 🙂 I was just looking through your comment on https://github.com/smallstep/crypto/pull/260, and saw the the link to your draft RFC, which in turn linked to the TPMKEY URI draft. That draft did inspire parts of our TPMKMS implementation, as well as the underlying packages. We chose to start with `tpmkms:`, because that fit our current implementation (and underlying packages) and we wouldn't be locked to the semantics of the draft RFC. 

We're close to making `step-tpm-plugin` public, which has more TPM specific functionality than this plugin. There's some functionality to make AKs and keys created by this KMS plugin and the TPM plugin be interoperable with the `tpm2_tools`, for example. It sounds like the file format you describe could also be supported through that plugin. The simplest would probably be to support exports in that format, similar to what we do for the `tpm2_tools` interoperability. 

We currently default to storing the serialized TPM objects (and corresponding certificates) as files in a directory. The files themselves contain a JSON representation of the TPM objects, which includes the serialized TPM data, as well as some metadata. It's a simple and effective solution and we don't need to maintain an additional (meta)data store that way. For the PEM format `BEGIN TSS2 PRIVATE KEY` I think we would need to keep some (meta)data outside of the PEM file? Or we could try encoding that in a PEM header, assuming tooling that doesn't understand our header, ignores it? Our current storage backend does allow for extension, so I could give it a try.

hslatman commented on go.mod:
```diff
@@ -6,59 +6,69 @@ require (
 	github.com/spf13/cobra v1.7.0
 	github.com/spf13/pflag v1.0.5
-	go.step.sm/crypto v0.31.2
+	go.step.sm/crypto v0.31.3-0.20230608211721-e41114688a00
```
Yes, that's the plan 🙂 While testing this with an actual TPM I found some small issues to be fixed before releasing a new version.

hslatman replied:
@maraino the PR at https://github.com/smallstep/crypto/pull/261 is ready to be reviewed. After merging that, I'll release a new `crypto` version and update this PR.

dwmw2 commented:
Note that the URI and the file format are fairly separate.

The URI would be used for keys which are stored *within* the TPM, in its NV storage. There was that old [tpmuri](https://datatracker.ietf.org/doc/html/draft-mavrogiannopoulos-tpmuri-01) draft from @nmav for TPMv1.2, and I've recently been [prodding](https://github.com/tpm2-software/tpm2-openssl/issues/71) the folks working on OpenSSL engines to get their act together and have a single form of identifier for TPMv2.0 too.

The file format is for keys which are stored on the file system (or anywhere else really) in encrypted form and *transiently* loaded into the TPM as required. The standardisation of that one is much further along; @jejb's [document](https://www.hansenpartnership.com/draft-bottomley-tpm2-keys.html) is new but the format is already supported in released versions of both OpenSSL engines as well as OpenConnect and GnuTLS.

Note that the file format does have [implications](https://gitlab.com/gnutls/gnutls/-/issues/594#note_688892934) about precisely how you generate the temporary parent key that wraps the stored key.

Being able to export to this PEM form seems like a reasonable choice. I believe tpm2_tools can use it too, so you could perhaps do it *instead* of the older separate pubkey/privkey blobs that tpm2_tools supported? Again, we do have to ensure the generated parent is compatible.

I think PEM headers should probably work, and you can also put stuff *outside* the `-----BEGIN…` … `-----END…` lines in the PEM file.

hslatman commented:
> Note that the URI and the file format are fairly separate.

Noted and understood!

> The URI would be used for keys which are stored _within_ the TPM, in its NV storage. There was that old [tpmuri](https://datatracker.ietf.org/doc/html/draft-mavrogiannopoulos-tpmuri-01) draft from @nmav for TPMv1.2, and I've recently been [prodding](https://github.com/tpm2-software/tpm2-openssl/issues/71) the folks working on OpenSSL engines to get their act together and have a single form of identifier for TPMv2.0 too.

That makes sense. Is it only for keys within the TPM, though? Because the old URI does support `tpmkey-file`? There's no explanation other than its name in the draft RFC, and it's just a draft, so maybe it wasn't all thought through nor commented on? 

> The file format is for keys which are stored on the file system (or anywhere else really) in encrypted form and _transiently_ loaded into the TPM as required. The standardisation of that one is much further along; @jejb's [document](https://www.hansenpartnership.com/draft-bottomley-tpm2-keys.html) is new but the format is already supported in released versions of both OpenSSL engines as well as OpenConnect and GnuTLS.

That looks great! I don't remember having seen that document before, but it does describe things I've been thinking about to store in our own serialization format, such as auth policies. The underlying Go libraries we depend on don't support passing these through currently, afaik, but work's being done that might make that possible. So this definitely looks interesting to support in the (near) future.

> Note that the file format does have [implications](https://gitlab.com/gnutls/gnutls/-/issues/594#note_688892934) about precisely how you generate the temporary parent key that wraps the stored key.

👍 

> Being able to export to this PEM form seems like a reasonable choice. I believe tpm2_tools can use it too, so you could perhaps do it _instead_ of the older separate pubkey/privkey blobs that tpm2_tools supported? Again, we do have to ensure the generated parent is compatible.

The `step-tpm-plugin` indeed does rely on the blobs, and I've been able to load the exported blobs with `tpm2_tools` again, so that's a start. Definitely sounds like something to explore 🙂 

> I think PEM headers should probably work, and you can also put stuff _outside_ the `-----BEGIN…` … `-----END…` lines in the PEM file.

Headers feel a bit neater to me, but that additional data part at the end might be a nice alternative. We do have some utilities for PEM deser thay may, or may not support the latter, but we can probably add support for that. The Go stdlib seems to do the right thing already.

dwmw2 commented:
> That makes sense. Is it only for keys within the TPM, though? Because the old URI does support tpmkey-file? There's no explanation other than its name in the draft RFC, and it's just a draft, so maybe it wasn't all thought through nor commented on?

`<rant>`
Yeah, I don't think `tpmkey-file` should have been in that draft; it was a bad idea. A file is a file. Users may have all kinds of stuff in files. PKCS#1, PKCS#8, PKCS#12 are all ways of storing actual private keys, it could be a TPMv1.2 TCPA blob wrapped in `-----BEGIN TSS KEY BLOB-----`, it could be a TPMv2.0 key in the form from the Bottomley draft, and probably others.

I object *vehemently* to asking the user to *tell* the application what kind of thing is stored in the file. The application should just *look* in the file and Do The Right Thing. 

I *hate* the fact that OpenSSL has *different* APIs for opening different types of PKCS#(1|8|12) files, even for PEM vs. DER versions of the *same* format, which leads to tools like `curl` explicitly pushing that as an "option" to the user instead of doing it right. It's awful.

So asking the user to specify `tpmkey:file=foo.pem` instead of `foo.pem` is just outright wrong. We should only do that if we *hate* our users. Which we don't.
`</rant>`

Note that the documentation of the TPMv2 format explicitly notes that it starts with an unambiguous OID specifically so that it can be detected even in DER form.

OpenConnect will Just Work™ with whatever file it's given, whether it be a software key in one of the myriad of formats, a TPMv1 TCPA key blob, a TPMv2 file. Or if it isn't a file, and is instead a PKCS#11 URI of the form `pkcs11:manufacturer=piv_II;id=%01`. It all Just Works.

hslatman commented:
@maraino I added support for creating a new key when running `attest` because of how key attestation works with the optional `QualifyingData` input, which depends on the ACME key authorization. It can't be read from `stdin` yet, but that's probably the next step. It does require creating a new key at the time of running `attest`, so `--new` is required. The parameters from `name` will be taken, so `tpmkms:name=new-key;ak=my-ak;qualifying-data=<hex>` will work.

The previous method, with the AK and key being created in separate steps, also still works.

I'm trying to fix RSA >2048 bits keys now, which is why I put the default to that. Apparently I've been testing with 2048 all the time; this was used all over the place in `go-tpm` and `go-attestation` 😅  The error seems to originate from the TPM (`RCValue`) and I see it happening with the simulator as well as my hardware TPM. Curiously, `tpm2_tools` reports the same error when trying `-G rsa3072:rsaes`, for example. Maybe my TPM and the simulator both don't support it.

maraino commented on cmd/attest.go:
```diff
@@ -230,7 +299,18 @@ func init() {
+	format := flagutil.LowerValue("format", []string{"", "step", "packed", "tpm"}, "")
+	flags.Var(format, "format", "The `format` to print the attestation.\nOptions are step, packed or tpm")
+	flags.Bool("leaf", false, "Print only the leaf certificate in a chain")
+	flags.Bool("new", false, "(EXPERIMENTAL) Creates and attests a new key instead of attesting an existing one")
+	flags.Var(kty, "kty", "The key `type` to build the certificate upon.\nOptions are EC and RSA")
+	flags.Var(crv, "crv", "The elliptic `curve` to use for EC and OKP key types.\nOptions are P256, P384 and P521")
+	flags.Int("size", 2048, "The key size for an RSA key") // TODO(hs): attesting 3072 bit RSA keys on TPM that doesn't support it returns an ugly error; we want to catch that earlier.
+	flags.Var(alg, "alg", "The hashing `algorithm` to use on RSA PKCS #1 and RSA-PSS signatures.\nOptions are SHA256, SHA384 or SHA512")
```
Probably mentioning that this is for tpmkms only or just for new keys. You're also using always PKCS#1 so the RSA-PSS in the help seems unnecessary. We can adding it back if we add support for --new on other KMSs.

maraino reviewed (commented):
I'm still unconvinced about the --bundle. I've added my reasoning in a comment.

hslatman commented:
> I'm still unconvinced about the --bundle. I've added my reasoning in a comment.

I don't see how I changed the existing logic for the `certificate` command. There used to be a `ReadCertificate`, `LoadCertificate` and `StoreCertificate`. These all operate on a single `*x509.Certificate`. The `CertFS` also operates on a single certificate, through the `CertificateManager` interface. As far as I know, it always printed only a single certificate when a single name/kuri was provided. Then `cryptoutil.LoadCertificateBundle` in the CLI would always get a single certificate.

I've changed `ReadCertficate` to `ReadCertificateBundle`. The first cert in the bundle has become the `cert` from the original call to `ReadCertificate`. Only when the KMS supports the `CertificateChainManager` will the chain methods be used (which can be used with a single cert); it'll use the `CertificateManager` interface otherwise. Only when `--bundle` is provided is the logic changed for outputting the certificate (chain), but since a single cert was supported only, that's functionally not different. The `CertFS` case is a special one that I couldn't readily adopt with the `CertificateChainManager` without refactoring a bit more than I was willing to do right now. It should still be the same logic as before adding `--bundle`.

Is it possible you were testing this with SoftKMS, with an empty `kms` in the CLI? Because that's a special case that would return a bundle.

dwmw2 commented on cmd/key.go:
```diff
@@ -55,17 +55,30 @@ var keyCmd = &cobra.Command{
+
+  # Get a key from the default TPM KMS:
+  step-kms-plugin key tpmkms:name=my-key
+  
```
Can we have a way to *export* the key to a `TSS2 PRIVATE KEY` PEM file, please? And to import them too.

hslatman replied:
I've added an issue for this: https://github.com/smallstep/step-kms-plugin/issues/73. I'm in favor of adding support, but I have a couple other high(er) priority things on my plate to finish first.

maraino reviewed (approved):
lgtm

dwmw2 commented:
> > I think PEM headers should probably work, and you can also put stuff _outside_ the `-----BEGIN…` … `-----END…` lines in the PEM file.
> 
> Headers feel a bit neater to me, but that additional data part at the end might be a nice alternative. We do have some utilities for PEM deser thay may, or may not support the latter, but we can probably add support for that. The Go stdlib seems to do the right thing already. One thing that our current format does, is storing any certificate (chains) belonging to an AK or key in the same file as as the serialized key data; with this PEM format, we may want to have a way to separate those into two files depending on how systems using the file want them to look like.

I stumbled across [RFC7468](https://datatracker.ietf.org/doc/html/rfc7468#page-4) today and saw this (penultimate para of p4):
```
   Unlike legacy PEM encoding [RFC1421], OpenPGP ASCII armor, and the
   OpenSSH key file format, textual encoding does *not* define or permit
   headers to be encoded alongside the data.  Empty space can appear
   between the pre-encapsulation boundary and the base64, but generators
   SHOULD NOT emit such any such spacing.  (The provision for this empty
   area is a throwback to PEM, which defined an "encapsulated header
   portion".)
```
